test images: Adds Windows Container images support (part 1)

claudiubelu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature

/sig testing
/sig windows

**What this PR does / why we need it**:

Adds Windows support to the ``test/images/image-util.sh`` script.

Adds support for building test images for multiple Windows versions, as we have to support both LTS and SAC channels. In order to build Windows container images for multiple OS versions, ``--isolation=hyperv`` is required. However, not all clouds / nodes supports or have it enabled by default, which is why we're going to rely on having multiple nodes to build the Windows images, until this issue is addressed.

All Windows images will be based on the images: 
``mcr.microsoft.com/windows/servercore:ltsc2019``, `mcr.microsoft.com/windows/servercore:1903`, `mcr.microsoft.com/windows/servercore:1909`). There are a couple of features that are needed from this image, especially powershell.

A Windows node with Docker installed is required to build Windows images. The connection URL to it must be set in the ``REMOTE_DOCKER_URL_${os_version}`` env variable. Additionally, the authentication to the remote docker node is done through certificates, which must be found in ``~/.docker-${os-version}``.

By default, the ``REMOTE_DOCKER_URL_${os_version}`` env variable is set to ``""`` in the Makefile, and because of it, the ``image-util.sh`` script will skip building and pushing Windows images.

Added ``GOOS`` argument to the go build process in order to be able to build Windows binaries. Additionally, the ``OS`` env variable was added to the images Makefiles (default value is ``linux``) in order to maintain default behaviour.

Some images require a different ``Dockerfile`` for Windows images, since they have different ways of  installing dependencies. Because of this, if a image needs to be built for Windows, it will first check for a ``Dockerfile_windows`` file instead of the default one. If there isn't one, it means that the same ``Dockerfile``can be used for both Windows and Linux.

Added ``busybox`` image for Windows. Most Windows images will be based on it, which will help reduce the command line differences between Linux and Windows, but not entirely.

Adds a README explaining the image building process, including the Windows Container image building process.

Changes the image naming template from:

    $REGISTRY/$IMAGE-$arch:$TAG

to

    $REGISTRY/$IMAGE:$TAG-$os_name-$arch

The previous naming template would generate a plethora of images (``Ai * N`` images, where Ai is the number of OS/architectures for the image i and N is the number of images), while the new naming template will reduce the number of images to N.

The new template also includes the OS name, as we plan to integrate Windows images into the manifest lists as well.

When building images, their ``REGISTRY`` can be set to a custom one, instead of the default ``gcr.io/kubernetes-e2e-test-images``. Some images are based on other images we're already building
(e.g.: ``kitten``, ``nautilus``), but their base images are set in the default registry name, which can be undesirable. This PR will allow the ``image-util.sh`` script to configure the ``REGISTRY`` for those images.

Windows images will require other base images, and thus, we will need to explicitly specify the OS type a base image is for in order to avoid confusion or errors. The ``BASEIMAGE`` entries now have an OS prefix.

Bumps the image versions in order to avoid breaking any currently built images and breaking the CIs.

Enables the Image Promoter to use Windows build nodes to build the Windows test images.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #77268

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

claudiubelu committed:
images: Adds linux/ prefix to BASEIMAGE entries

Windows images will require other base images, and thus, we will need
to explicitly specify the OS type a base image is for in order to
avoid confusion or errors.

claudiubelu committed:
test images: Adds Windows support (part 1)

Adds Windows support to the test/images/image-util.sh script.

A Windows node with Docker installed is required to build Windows images.
The connection URL to it must be set in the REMOTE_DOCKER_URL env variable.
Additionally, the authentication to the remote docker node is done through
certificates, which must be found in ~/.docker.

By default, the REMOTE_DOCKER_URL env variable is set to "" in the Makefile,
and because of it, the image-util.sh script will skip building and pushing
Windows images.

Added GOOS argument to the go build process in order to be able to build
Windows binaries. Additionally, the OS env variable was added to the images
Makefiles (default value is "linux") in order to maintain default behaviour.

Some images require a different Dockerfile for Windows images, since they
have different ways of installing dependencies. Because of this, if a image
needs to be built for Windows, it will first check for a Dockerfile_windows
file instead of the default one. If there isn't one, it means that the
same Dockerfile can be used for both Windows and Linux.

All Windows images will be based on the image
"mcr.microsoft.com/windows/servercore:ltsc2019". There are a couple of features
that are needed from this image, especially powershell.

Added busybox image for Windows. Most Windows images will be based on it, which
will help reduce the command line differences between Linux and Windows, but
not entirely.

Added Windows support for agnhost image.

claudiubelu committed:
images: Changes the image naming template

Changes the image naming template from:

$REGISTRY/$image-$arch:$TAG

to

$REGISTRY/$image:$TAG-$os_name-$arch

The previous naming template would generate a plethora of images (Ai * N images,
where Ai is the number of OS/architectures for the image i and N is the number
of images), while the new naming template will reduce the number of images to N.

The new template also includes the OS name, as we plan to integrate Windows
images into the manifest lists as well.

claudiubelu committed:
images: Configurable BASEIMAGE hierarchy

When building images, their REGISTRY can be set to a custom
one, instead of the default "gcr.io/kubernetes-e2e-test-images" or
"us.gcr.io/k8s-artifacts-prod/e2e-test-images".
Some images are based on other images we're already building
(e.g.: kitten, nautilus), but their base images
are set in the default registry name, which can be undesirable.

This commit addresses this issue.

neolit123 commented:
/assign @mkumatag
/priority important-longterm

neolit123 commented on test/images/resource-consumer/consume-cpu/consume_cpu_windows.go:
```diff
@@ -0,0 +1,92 @@
+	}
+}
+
+type ProcCPUStats struct {
```
please add code comments to pass `go lint` on the new go files.

bclau commented:
/retest

bclau commented:
/retest

adelina-t commented:
/cc @PatrickLang

PatrickLang commented:
cc @dims

dims commented:
/assign @listx

mkumatag commented on test/images/image-util.sh:
```diff
@@ -96,8 +112,16 @@ build() {
+      # ca.pem, cert.pem, key.pem files can be found in the ~/.docker folder.
+      docker --tlsverify -H "${REMOTE_DOCKER_URL}" build --pull -t "${REGISTRY}/${IMAGE}:${TAG}-${os_name}-${arch}" -f $dockerfile_name .
+    else
+      echo "Cannot build the image '${IMAGE}' for ${os_name}/${arch}. REMOTE_DOCKER_URL should be set, containing the URL to a Windows docker daemon."
```
are you considering these windows build as optional if REMOTE_DOCKER_URL is not set?

bclau replied:
Yep. Otherwise it would break any current building processes.

mkumatag commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
+      os_name=`echo $base_image | cut -d "/" -f 1`
+      arch=`echo $base_image | cut -d "/" -f 2`
+    fi
+    docker manifest annotate --arch "${arch}" "${REGISTRY}/${IMAGE}:${TAG}" "${REGISTRY}/${IMAGE}:${TAG}-${os_name}-${arch}"
```
we may need to pass `--os` explicitly as we are building it for windows os as well!

Help:
```
╰─ docker manifest annotate --help

Usage:	docker manifest annotate [OPTIONS] MANIFEST_LIST MANIFEST

Add additional information to a local image manifest

Options:
      --arch string           Set architecture
      --os string             Set operating system
      --os-features strings   Set operating system feature
      --variant string        Set architecture variant
```

bclau replied:
True, I forgot to add that.

mkumatag reviewed (changes requested):
First of all thanks for this PR. Just got few minutes to review and here are my review comments, will take some more time later to review in detail.

One very basic question I have is: who will build the windows images on that remote machine? Will that machine will be accessible for whoever is building/pushing images now.

mkumatag commented on test/images/dnsutils/Dockerfile_windows:
```diff
@@ -0,0 +1,23 @@
+# Copyright 2019 The Kubernetes Authors.
```
In few places I see that `dnsmasq` command has been called part of pod, not sure that case is handled here! - https://github.com/kubernetes/kubernetes/blob/2a05e1b76f9f0fa52a31701ba0074d8d431da2ec/test/e2e/network/dns_common.go#L292

bclau replied:
True, and it is not handled. I haven't found a Windows equivalent for that yet, but this will be sufficient for most DNS tests to pass.

claudiubelu reviewed (commented):
> 
> 
> First of all thanks for this PR. Just got few minutes to review and here are my review comments, will take some more time later to review in detail.
> 
> One very basic question I have is: who will build the windows images on that remote machine? Will that machine will be accessible for whoever is building/pushing images now.

Are you speaking conceptually, or in practice? Basically, through this PR, the Linux build node will be able to send the docker build / push commands remotely to the Windows build node.

The goal is to also have Windows images built in the same process that builds and publishes the images to ``gcr.io/kubernetes-e2e-test-images``. AFAIK, only Google folks have the rights to push those images there, so they will have to also create a Windows build node that is able to push to that repository.

PatrickLang commented on test/images/README.md:
```diff
@@ -0,0 +1,113 @@
+# Building Kubernetes test images
```
:) docs are good

PatrickLang commented on test/images/apparmor-loader/BASEIMAGE:
```diff
@@ -1,4 +1,4 @@
-arm64=arm64v8/alpine:3.8
-ppc64le=ppc64le/alpine:3.8
+linux/amd64=alpine:3.8
+linux/arm=arm32v6/alpine:3.8
```
I thought it was armhf?

dims replied:
@PatrickLang we have been using this for a while now - https://cs.k8s.io/?q=arm32v6&i=nope&files=&repos=kubernetes/kubernetes

timothysc commented:
/assign @dims @johnSchnake

bclau commented:
/test pull-kubernetes-integration

johnSchnake commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
+    # NOTE(claudiub): if the REMOTE_DOCKER_URL var is not set, or it is an empty string, we must skip
+    # pushing the Windows image and including it into the manifest list.
+    if [[ ! -v "REMOTE_DOCKER_URL" || -z "${REMOTE_DOCKER_URL}" ]]; then
+      if [[ ! -z "$(printf "%s\n" $base_images | grep '^windows')" ]]; then
```
If (pun intended) you can find a way to less deeply nest these conditions I think it might find it easier to follow. Right now the there is a path with no extra processing (if `REMOTE_DOCKER_URL` isn't set but it doesn't include windows in the base images.

Seems like these 2 `ifs` you're adding are really just one longer condition rather than wanting to separately handle up to 4 logical paths (if/else on each).

listx replied:
Why not `-n` instead of `! -z` ?

listx replied:
+1 to unnest/combine these 2 `ifs`

johnSchnake commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
+      os_name=`echo $base_image | cut -d "/" -f 1`
+      arch=`echo $base_image | cut -d "/" -f 2`
+    fi
+    docker manifest annotate --os "${os_name}" --arch "${arch}" "${REGISTRY}/${IMAGE}:${TAG}" "${REGISTRY}/${IMAGE}:${TAG}-${os_name}-${arch}"
```
> The previous naming template would generate a plethora of images (Ai * N images, where Ai is the number of OS/architectures for the image i and N is the number of images), while the new naming template will reduce the number of images to N.

Could you clarify the intent a bit on this? In the end you still have `#os X #arch X #tag` number of images/tags. Before you'd have `img-windows/amd64:tag` and now you have `img:tag-windows-amd64`. In either case docker will be reutilzing the layers so isn't the end result the same?

bclau commented on test/images/resource-consumer/consume-memory/consume_memory.go:
```diff
@@ -0,0 +1,99 @@
+	flag.Parse()
+	sizeB, err = parseSize(*sizestr)
+	if err != nil {
+		fmt.Println("got error: ", err)
```
True. I could even ``Fatalf``, which does both.

bclau commented on test/images/resource-consumer/consume-cpu/consume_cpu_windows.go:
```diff
@@ -0,0 +1,92 @@
+func main() {
+	phandle, err := syswin.GetCurrentProcess()
+	if err != nil {
+		panic(err)
```
Done

bclau commented:
I wonder if we should bump the image versions for the affected images in order to avoid any potential future failures in the near future. That will make it easier to just fallback to the current stable image versions.

claudiubelu committed:
test images: Adds building README

Adds a README explaining the image building process, including the
Windows Container image building process.

claudiubelu committed:
test images: Bumps image versions

The way the images are built is going to be changed, and in order to avoid
overwritting and breaking the current images, the image versions are bumped.

listx commented:
> I wonder if we should bump the image versions for the affected images in order to avoid any potential failures in the near future. That will make it easier to just fallback to the current stable image versions.

I think bumping the versions would help.

Is there any way to break up this PR into smaller, independent pieces? I ask because I don't feel confident about reviewing such a large PR. I think you could at least break it up into 2, by separating the enable-building-windows-images logic from the actual addition of the Dockerfile_windows files that make use of the new windows build logic.

bclau commented:
> 
> 
> > I wonder if we should bump the image versions for the affected images in order to avoid any potential failures in the near future. That will make it easier to just fallback to the current stable image versions.
> 
> I think bumping the versions would help.
> 

Then I'll include a commit which bumps the versions in the split PR.

> Is there any way to break up this PR into smaller, independent pieces? I ask because I don't feel confident about reviewing such a large PR. I think you could at least break it up into 2, by separating the enable-building-windows-images logic from the actual addition of the Dockerfile_windows files that make use of the new windows build logic.

Will do. I've left here the Windows support for the script and for the busybox and agnhost image. The rest will be included in another PR.

listx commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
+    base_images=$(listBaseImages)
+    # NOTE(claudiub): if the REMOTE_DOCKER_URL var is not set, or it is an empty string, we must skip
+    # pushing the Windows image and including it into the manifest list.
+    if [[ ! -v "REMOTE_DOCKER_URL" || -z "${REMOTE_DOCKER_URL}" ]]; then
```
Dumb question: does the `!` here mean (in pseudo bash)

```
 if [[ ! (-v "REMOTE_DOCKER_URL") || -z "${REMOTE_DOCKER_URL}" ]]
```

or

```
 if [[ ! (-v "REMOTE_DOCKER_URL" || -z "${REMOTE_DOCKER_URL}") ]]
```
?

listx replied:
Dumb question: does the `!` only apply to the first half of the `... || ...` or to the entire expression?

listx commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
+    if [[ ! -v "REMOTE_DOCKER_URL" || -z "${REMOTE_DOCKER_URL}" ]]; then
+      if [[ ! -z "$(printf "%s\n" $base_images | grep '^windows')" ]]; then
+        echo "Skipping pushing the image '${IMAGE}' for Windows. REMOTE_DOCKER_URL should be set, containing the URL to a Windows docker daemon."
+        base_images=$(printf "%s\n" $base_images | grep -v "^windows")
```
Why not move this assignment above line 148 so that you don't have to re-type the same logic here as inside the conditional in 148?

bclau replied:
It looks similar, but it is not. The conditional greps ^windows to check if we have to build the image for Windows, while here it is grep -v, which excludes all lines with ^windows.

listx commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
   else
-    archs=${!QEMUARCHS[@]}
+    # prepend linux/ to the QEMUARCHS items.
+    base_images=`printf 'linux/%s\n' "${!QEMUARCHS[@]}"`
```
Not sure why you chose backticks here over `$(...)`

bclau replied:
Personal preference, honestly, but I do try to keep it consistent with how the file is written. Seems like I forgot about this.

listx commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
+    if [[ $base_image =~ .*/.* ]]; then
+      os_name=`echo $base_image | cut -d "/" -f 1`
+      arch=`echo $base_image | cut -d "/" -f 2`
+    fi
```
Can you handle the missing `else` condition here?

bclau replied:
There's no ``else`` case. There are 2 cases when building an image: 

- it has  ``BASEIMAGE`` file, Which this PR updates them so **all** of them are os/arch formatted. Also, the ``getBaseImage`` now requires both OS and arch. The grep will fail without both of them now.
- ``BASEIMAGE`` does not exist, in which case ``linux/`` is prepended above on L51.

listx replied:
I think it's always a good idea to handle `else` statements for every `if`, no matter what the surrounding state is *supposed* to be (at the very least, you can at least print an error message and exit with an error).

bclau replied:
Ok, added ``else``.

listx commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
+    if [[ $base_image =~ .*/.* ]]; then
+      os_name=`echo $base_image | cut -d "/" -f 1`
+      arch=`echo $base_image | cut -d "/" -f 2`
+    fi
```
Please handle the missing `else` condition

bclau replied:
There's no ``else`` case. There are 2 cases when building an image: 

- it has  ``BASEIMAGE`` file, Which this PR updates them so **all** of them are os/arch formatted. Also, the ``getBaseImage`` now requires both OS and arch. The grep will fail without both of them now.
- ``BASEIMAGE`` does not exist, in which case ``linux/`` is prepended above on L51.

listx commented on test/images/image-util.sh:
```diff
@@ -44,15 +45,21 @@ getBaseImage() {
+    if [[ $base_image =~ .*/.* ]]; then
+      os_name=`echo $base_image | cut -d "/" -f 1`
+      arch=`echo $base_image | cut -d "/" -f 2`
+    fi
```
Please handle the missing `else` case

listx commented on test/images/image-util.sh:
```diff
@@ -146,7 +193,7 @@ bin() {
         /bin/bash -c "\
                 cd /go/src/k8s.io/kubernetes/test/images/${SRC_DIR} && \
-                CGO_ENABLED=0 GOARM=${GOARM} GOARCH=${ARCH} go build -a -installsuffix cgo --ldflags '-w' -o ${TARGET}/${SRC} ./$(dirname "${SRC}")"
+                CGO_ENABLED=0 GOARM=${GOARM} GOOS=${OS} GOARCH=${ARCH} go build -a -installsuffix cgo --ldflags '-w' -o ${TARGET}/${SRC} ./$(dirname "${SRC}")"
```
Should this be `${GOOS}` instead of `${OS}`?

listx commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
+    if [[ ! -v "REMOTE_DOCKER_URL" || -z "${REMOTE_DOCKER_URL}" ]]; then
+      if [[ ! -z "$(printf "%s\n" $base_images | grep '^windows')" ]]; then
+        echo "Skipping pushing the image '${IMAGE}' for Windows. REMOTE_DOCKER_URL should be set, containing the URL to a Windows docker daemon."
+        base_images=$(printf "%s\n" $base_images | grep -v "^windows")
```
This line looks awfully similar to the conditional inside line 148. Is there a way to somehow make it more DRY?

bclau replied:
It looks similar, but it is not. The conditional greps ``^windows`` to check if we have to build the image for Windows, while here it is ``grep -v``, which excludes all lines with ``^windows``.

listx commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
+    if [[ $base_image =~ .*/.* ]]; then
+      os_name=`echo $base_image | cut -d "/" -f 1`
+      arch=`echo $base_image | cut -d "/" -f 2`
+    fi
```
Can you handle the missing `else`' condition here?

listx commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
+    if [[ $base_image =~ .*/.* ]]; then
+      os_name=`echo $base_image | cut -d "/" -f 1`
+      arch=`echo $base_image | cut -d "/" -f 2`
+    fi
```
Missing `else` condition?

listx commented:
Oops, sorry for double-posting in the code review. Looks like I forgot that I pressed the submit-button.

listx commented on test/images/image-util.sh:
```diff
@@ -117,23 +141,46 @@ push() {
-  # Make archs list into image manifest. Eg: 'amd64 ppc64le' to '${REGISTRY}/${IMAGE}-amd64:${TAG} ${REGISTRY}/${IMAGE}-ppc64le:${TAG}'
-  manifest=$(echo "$archs" | ${SED} -e "s~[^ ]*~$REGISTRY\/$IMAGE\-&:$TAG~g")
+  # Make base_images list into image manifest. Eg: 'linux/amd64 linux/ppc64le' to '${REGISTRY}/${IMAGE}:${TAG}-linux-amd64 ${REGISTRY}/${IMAGE}:${TAG}-linux-ppc64le'
+  manifest=$(echo "$base_images" | ${SED} "s~\/~-~g" | ${SED} -e "s~[^ ]*~$REGISTRY\/$IMAGE:$TAG\-&~g")
```
If the BASEIMAGE files are formatted as

```
os-arch=some/image:tag
```

instead of

```
os/arch=some/image:tag
```

(don't use `/`), does that allow us to simplify the call to to `sed` here (notably, drop the middle one to `${SED} "s~\/~-~g"`? I think because of the `g` it might actually be a bug because you are replacing all instances of `/` which includes the ones possibly embedded in the image names.

Apologies if I am misinterpreting this code.

bclau replied:
Indeed, I just realized, the names ``listBaseImages`` and ``base_images`` are misleading and they should be renamed. In actuality, the command that is executed is:

    cut -d "=" -f 1 "${IMAGE}"/BASEIMAGE

which basically results in:

    linux/amd64
    linux/arm
    linux/arm64
    linux/ppc64le
    linux/s390x
    windows/amd64

Indeed, we can drop the ``g`` to be safe, but it won't change anything.

listx commented:
/lgtm

mkumatag commented on test/images/README.md:
````diff
@@ -0,0 +1,113 @@
+
+```bash
+    REMOTE_DOCKER_URL=remote_docker_url REGISTRY=foo_registry make all-push WHAT=test-webserver
+```
````
Please add the issue mentioned under known issue and workaround from - https://github.com/docker/for-linux/issues/396 to make doc complete

bclau replied:
Done, I've simply added the issue title and a link to it. All the information and solutions can be read on the issue, and if the issue will ever be reopened, or if there will be new information, it will probably be added there.

mkumatag replied:
thanks for adding that information @bclau

mkumatag commented:
Changes look okay to me.

@listx @ixdy help here to push all those images with a new tag and later [manifest.go](https://github.com/kubernetes/kubernetes/blob/master/test/utils/image/manifest.go) can be updated with those new versions and pass the tests?

bclau commented:
> 
> 
> Changes look okay to me.
> 
> @listx @ixdy help here to push all those images with a new tag and later [manifest.go](https://github.com/kubernetes/kubernetes/blob/master/test/utils/image/manifest.go) can be updated with those new versions and pass the tests?

This is just part 1, there is still part 2 and 3. But afterwards, yes, the images will have to be built and published and the [manifest.go](https://github.com/kubernetes/kubernetes/blob/master/test/utils/image/manifest.go) will have to be updated afterwards.

A Windows node will be required though. I could share mine if needed.

bclau commented:
/retest pull-kubernetes-bazel-build

bclau commented:
/test pull-kubernetes-bazel-build

spiffxp commented:
/cc

johnSchnake commented:
What is the status of this?

There are another series of PRs (one of them here: https://github.com/kubernetes/kubernetes/pull/79142) which is trying to centralize logic into the agnhost image.

Any idea how that will impact this? It may be unnecessary to be doing much of this work if so many fewer images need updated.

bclau commented:
> 
> 
> What is the status of this?
> 
> There are another series of PRs (one of them here: #79142) which is trying to centralize logic into the agnhost image.
> 
> Any idea how that will impact this? It may be unnecessary to be doing much of this work if so many fewer images need updated.

This PR series has been done for quite a while, and especially part 1 received quite a lot of reviews. I would get this in, that way it will allow me to also update the ``agnhost``'s ``Dockerfile_windows`` file.

Part 1 doesn't really conflict much with the PR series you mentioned, only part 2.  Plus, it's only the BASEIMAGE and VERSION files which would conflict anyways.

bclau commented:
/test pull-kubernetes-bazel-test

johnSchnake commented on test/images/agnhost/VERSION:
```diff
@@ -1 +1 @@
-2.1
+2.2
```
https://github.com/kubernetes/kubernetes/pull/79142 is specifically bumping this version as well; coordination needs to happen to ensure the versions/merges happen in the appropriate order.

bclau replied:
Correct. But if I'm not mistaken, a merge conflict will occur once one of the PR merges anyways.

spiffxp replied:
I'm going to assume we'll want to bump this to 2.3, and am fine aiming for all of the other agnhost PR's before going after the windows stuff

spiffxp commented:
/hold
for https://github.com/kubernetes/kubernetes/pull/79142 and https://github.com/kubernetes/kubernetes/pull/79238 to land

bclau commented:
Rebased and since we decided to merge this after the Centralization part 3 PR, I've removed any changes to those images, since it will no longer be necessary. Additionally, I've added the things to ``agnhost/Dockerfile_windows`` related to those images from part 3 (iperf, entrypoint-tester, porter). Also added a few more details to ``agnhost/README.md`` regarding the Windows image.

spiffxp commented on test/images/image-util.sh:
```diff
@@ -117,23 +144,50 @@ push() {
+      docker push "${REGISTRY}/${IMAGE}:${TAG}-${os_name}-${arch}"
+    else
+      # NOTE(claudiub): We're pushing the image we built on the remote Windows node.
+      docker --tlsverify -H "${REMOTE_DOCKER_URL}" push "${REGISTRY}/${IMAGE}:${TAG}-${os_name}-${arch}"
```
I think we want to gate this behind the same check you're using for `build` above, ie:
```sh
elif [[ -v "REMOTE_DOCKER_URL" && ! -z "${REMOTE_DOCKER_URL}" ]]; then
#...
else
  echo "Cannot push ..."
fi
```

bclau replied:
Sure, but it shouldn't be the case in the first place. At L150, if ``$REMOTE_DOCKER_URL`` is not set, we filter out the ``windows/`` entries from ``os_archs``, so the ``elif`` branch will never be executed.

spiffxp commented on test/images/image-util.sh:
```diff
@@ -96,8 +115,16 @@ build() {
-
+    if [[ "$os_name" = "linux" ]]; then
+      docker build --pull -t "${REGISTRY}/${IMAGE}:${TAG}-${os_name}-${arch}" .
+    elif [[ -v "REMOTE_DOCKER_URL" && ! -z "${REMOTE_DOCKER_URL}" ]]; then
```
```suggestion
    elif [[ -n "${REMOTE_DOCKER_URL:-}" ]]
```

spiffxp commented on test/images/Makefile:
```diff
@@ -15,6 +15,7 @@
 include  ../../hack/make-rules/Makefile.manifest
 
 REGISTRY ?= gcr.io/kubernetes-e2e-test-images
+REMOTE_DOCKER_URL ?= ""
```
This was being set to `""` instead of the empty string
```suggestion
REMOTE_DOCKER_URL ?=
```

spiffxp commented on test/images/image-util.sh:
```diff
@@ -117,23 +144,50 @@ push() {
+    os_archs=$(listOsArchs)
+    # NOTE(claudiub): if the REMOTE_DOCKER_URL var is not set, or it is an empty string, we must skip
+    # pushing the Windows image and including it into the manifest list.
+    if [[ ((! -v "REMOTE_DOCKER_URL") || -z "${REMOTE_DOCKER_URL}") && -n "$(printf "%s\n" $os_archs | grep '^windows')" ]]; then
```
```suggestion
    if [[ -z "${REMOTE_DOCKER_URL:-}" && -n "$(printf "%s\n" $os_archs | grep '^windows')" ]]; then

```

spiffxp reviewed (commented):
Bumped into some things trying this locally

dims commented:
/uncc

spiffxp reviewed (commented):
Stuck on `jessie-dnsutils`

bclau commented on test/images/jessie-dnsutils/VERSION:
```diff
@@ -1 +1 @@
-1.1
```
There is a reason why jessie-dnsutils and someone knows it. I have heard it once, but I don't remember what it was, but I think it was related to the base images and potential behaviours they could have.

spiffxp replied:
ok I agree https://github.com/kubernetes/kubernetes/pull/80053 will fix this, please rebase once that's merged

bclau commented:
/retest

bclau commented:
/retest

spiffxp commented:
I don't have bandwidth to examine why `jessie-dnsutils` is required, and consider this blocked until someone can either: explain how to build arm64 images for `jessie`, or explain why `jessie` was required and whether `stretch` could be used instead.

bclau commented:
> 
> 
> I don't have bandwidth to examine why `jessie-dnsutils` is required, and consider this blocked until someone can either: explain how to build arm64 images for `jessie`, or explain why `jessie` was required and whether `stretch` could be used instead.

Indeed, I can see it too now. This works for me now: https://github.com/kubernetes/kubernetes/pull/80053

spiffxp commented on test/images/agnhost/VERSION:
```diff
@@ -1 +1 @@
-2.2
+2.3
```
bump to 2.5 ref: https://github.com/kubernetes/kubernetes/pull/79423#issuecomment-510526158

spiffxp reviewed (commented):
getting there

spiffxp commented:
>The goal is to also have Windows images built in the same process that builds and publishes the images to gcr.io/kubernetes-e2e-test-images. AFAIK, only Google folks have the rights to push those images there, so they will have to also create a Windows build node that is able to push to that repository.

This part, I don't have the bandwidth to tackle.
/cc @yujuhong

bclau commented:
Rebased after the jessie-dnsutils fix merged, bumped agnhost version to 2.5

yujuhong commented on test/images/README.md:
```diff
@@ -0,0 +1,129 @@
+
+In order to build the Windows container images, a node with Windows 10 or Windows Server 2019
+with the latest updates installed is required. The node will have to have Docker installed,
+preferably version 18.06.0 or newer.
```
Can you be more explicit about why this version is preferred?

bclau replied:
Hm, there's no particular reason why it has to be 18.06.0 or newer on the Windows nodes. 18.06.0 or newer was chosen for consistency with the requirements on the Linux node (see L11).

BenTheElder replied:
most build things are 19.03 by now

yujuhong commented on test/images/README.md:
```diff
@@ -0,0 +1,129 @@
+Note which version of Docker is installed on the Windows node: Enterprise Edition, or
+Community Edition. If it is EE, the container OS version and the host's OS version **must** match
+in order to start the containers. Currently, the Windows Containers are built based on the
+`mcr.microsoft.com/windows/servercore:ltsc2019` image.
```
Don't we also want to build images for newer Windows OS images from the semi-annual release channels?

bclau replied:
That's going to be more difficult,

Currently, we officially support  WS 1809 / 2019 containers, which we also use in tests (although 1803 are also fine, and 1903 should also be fine), and IMO, it's a reasonable version to test since it is an LTS.

Now, adding multiple OS versions to the image building process was something I've had in mind, but it's not quite as simple. The difficult part would be Windows node(s). Currently, there would be 2 ways this could be done:

- have multiple Windows nodes with different OS versions and docker EE installed build images for those versions (as mentioned in this paraghraph, currently the container OS version must match the host OS version in order start / build the container images). This would also be a bit more difficult to implement and use, you would have to somehow map hosts with what image versions to build.
- have a single Windows node with docker CE installed. This is what I have been using, I had a Windows 10 node for building images for both 1803 and 1809 and creating manifest lists containing both of them. Now, the disadvantage to this, would be the build time. It takes a lot longer to build an image using this method.

At one point in the future, docker EE will also support multiple container OS versions, like docker CE does, but with hopefully better build times, and when that happens, we can increase the docker version requirement to that and stick with a single Windows node which can build any image we need.

One thing we could probably do and if there's a need for it, would be to parameterize the base Windows container image, so, if anyone wants to build on top of another Windows OS version, they can just set that parameter in the make command.

yujuhong replied:
> Currently, we officially support WS 1809 / 2019 containers, which we also use in tests (although 1803 are also fine, and 1903 should also be fine), and IMO, it's a reasonable version to test since it is an LTS.

Hmm..my concern is that if sig-windows supports a new version of Windows image in the next release, this would still need to be updated to build and test both SAC and LTS images. Is my understanding incorrect?

> * have multiple Windows nodes with different OS versions and docker EE installed build images for those versions (as mentioned in this paraghraph, currently the container OS version must match the host OS version in order start / build the container images).

This may not be that big of problem if we build the images using GCE VMs @pjh 



> At one point in the future, docker EE will also support multiple container OS versions, like docker CE does, but with hopefully better build times, and when that happens, we can increase the docker version requirement to that and stick with a single Windows node which can build any image we need.

I'm not familiar with the constraints (or lack thereof) docker CE. How does Docker CE handle building images for multiple OS versions? Does that also mean it can start containers even though versions don't match (without having any issues)?

bclau replied:
> 
> 
> > Currently, we officially support WS 1809 / 2019 containers, which we also use in tests (although 1803 are also fine, and 1903 should also be fine), and IMO, it's a reasonable version to test since it is an LTS.
> 
> Hmm..my concern is that if sig-windows supports a new version of Windows image in the next release, this would still need to be updated to build and test both SAC and LTS images. Is my understanding incorrect?

Right, it would depend on sig-windows to decide what versions are officially supported. But that would not only imply that we'll have to build images for both versions, but we would also have to add CI jobs for those versions as well.

> 
> > * have multiple Windows nodes with different OS versions and docker EE installed build images for those versions (as mentioned in this paraghraph, currently the container OS version must match the host OS version in order start / build the container images).
> 
> This may not be that big of problem if we build the images using GCE VMs @pjh
> 
> > At one point in the future, docker EE will also support multiple container OS versions, like docker CE does, but with hopefully better build times, and when that happens, we can increase the docker version requirement to that and stick with a single Windows node which can build any image we need.
> 
> I'm not familiar with the constraints (or lack thereof) docker CE. How does Docker CE handle building images for multiple OS versions? Does that also mean it can start containers even though versions don't match (without having any issues)?

Yeah, you can start containers even though versions don't match, within some boundaries. At one point, I had to deliberately install the Windows updates on my Windows 10 node in order to be able to build the latest WS 2019 images.

yujuhong replied:
Maybe I missed something, but I don't see how sig-windows could escape supporting multiple versions (or supporting both SAC and LTSC). Do we not expect to build/integrate features in released only in SAC at all...? @PatrickLang 

Hm....so Docker CE has looser constraints, but is there guarantee that the containers (with mismatched versions) would run without any issue? Trying to understand where the difference comes from. Is there any documentation that I could read?

bclau replied:
Ok, so, I have a bit more information about this.

First of all, the reason why multiple image versions were supported on my Windows 10 node is because Hyper-V isolation is enabled by default on Windows client, while Windows Server uses process isolation by default. See the explanation here: https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-configuration-file (at the "Options for the runtime" section).

Apparently, I can do this on my Windows Server 2019 node without any issue:

    docker run --isolation=hyperv -d --name test-1803 --entrypoint ping microsoft/windowsservercore:1803 -t localhost
    db99b69779941dd55e9a2877817e1597ab845c87ea477873f0230d4e61f01938
    docker run --isolation=hyperv -d --name test-2019 --entrypoint ping mcr.microsoft.com/windows/servercore:ltsc2019 -t localhost
    c57b09fdbafb9c932a34d6729c6ec3a6bb2064c4c300acb50ff820900b0d58c5

Interestingly, I cannot do the same for 1903 images:

    docker run --isolation=hyperv -d --name test-2019 --entrypoint ping mcr.microsoft.com/windows/servercore:1903-amd64 -t localhost
    Unable to find image 'mcr.microsoft.com/windows/servercore:1903-amd64' locally 1903-amd64: Pulling from windows/servercore
    C:\Program Files\Docker\docker.exe: a Windows version 10.0.18362-based image is incompatible with a 10.0.17763 host.
    See 'C:\Program Files\Docker\docker.exe run --help'.

So, it seems that for newer images, a newer Windows Server node is still required.

https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/hyperv-container

bclau replied:
@yujuhong I've added a commit to this PR (https://github.com/kubernetes/kubernetes/pull/76838/commits/0c744f5d2054605c9c6af38a31815de1b11405eb) which adds support for building for multiple Windows OS versions. I've tried it out for the ``busybox`` version, for WS 1803 and 1809 (since I have a WS 1809 node). Here's the resulting manifest list: https://paste.ubuntu.com/p/q7dhswZcdc/

I've also updated the Windows section in the README to reflect these things.

Is there anything else we need for this PR?

PatrickLang replied:
Windows Server 1803 - no support. We're not testing that in Kubernetes and declared Windows Server 2019 stable.

Windows Server 2019 & Windows Server version 1809 - same thing, different branding. The containers are interchangeable.

Windows Server version 1903 - let's discuss in SIG. There's no testing for it yet, but it's feasible

yujuhong replied:
Ack. I think it'd be valuable to support SAC releases (1903, etc.) to support new features. I agree this should be determined by sig-windows, and was hoping that the doc can be more generic so that we could augment supported versions when needed.

yujuhong reviewed (commented):
> This part, I don't have the bandwidth to tackle.

How's this done today? Do we just have a person manually go through the process to build and push all the images?

javier-b-perez commented on test/images/agnhost/Dockerfile_windows:
```diff
@@ -0,0 +1,44 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM BASEIMAGE
```
instead of string/replace could we use ARG?

bclau replied:
Indeed, I prefered ARG instead of string/replace mechanism we have here. (See: https://github.com/kubernetes-sigs/windows-testing/blob/master/images/test-webserver/Dockerfile#L15). If people want it, I can do this, it'll make the Dockerfiles usable outside of the ``image-util.sh`` script.

bclau commented:
@yujuhong is there anything from your side blocking this PR?

claudiubelu committed:
test images: Adds multiple Windows channels support

This commit adds support for building test images for multiple
Windows versions, as we have to support both LTS and SAC channels.

With this, the format for Windows images in the BASEIMAGE files is:

OS/ARCH/OS_VERSION

Also adds --isolation-hyperv to the Windows docker build command, making sure
that container images for multiple OS versions can be built using the same
Windows node.

yujuhong commented:
> @yujuhong is there anything from your side blocking this PR?

I don't have problems with this PR (and thanks for writing it!)
My questions are mostly around 1) what windows versions should sig-windows support (to be discussed/determined by sig-windows), 2) who will follow the instructions to build and push the windows images. The instructions won't work for GCE, but I think we can discuss the logistics separately.

ttousai commented:
Hello! I'm a bug triage shadow for the 1.16 release cycle and considering this issue is tagged for 1.16, I'd like to check its status. Code freeze is starting on August 29th (about 3 days from now), which means that this PR must merged by then.

Do we still target this issue to be fixed for 1.16?

PatrickLang commented:
/retest

PatrickLang commented:
/milestone v1.17
Given that we still don't have the reviews needed and closure on how to maintain the build hosts, I don't think we'll be done today. I'll put this in the 1.17 backlog and we should push for it ASAP then

bclau commented:
/retest

bclau commented:
@yujuhong Can you check this PR again? In the last commit I've added support for multiple Windows nodes being used to build multiple versions of Windows test images. With that, are there any blockers left?

yujuhong commented:
> @yujuhong Can you check this PR again? In the last commit I've added support for multiple Windows nodes being used to build multiple versions of Windows test images. With that, are there any blockers left?

No blockers from me. @yliaog @pjh if you'd like to take a look.

yliaog commented on test/images/busybox/Makefile:
```diff
@@ -0,0 +1,27 @@
+# limitations under the License.
+
+SRCS=hostname
+OS ?= windows
```
why is the default 'windows'?

yliaog commented on test/images/image-util.sh:
```diff
@@ -117,23 +158,75 @@ push() {
+      suffix="$os_name-$arch"
+    else
+      echo "The BASEIMAGE file for the ${IMAGE} image is not properly formatted. Expected entries to start with 'os/arch', found '${os_arch}' instead."
+      exit 1
```
almost identical code in build() / push(), better to extract into a function?

bclau replied:
I don't think it will help much. Most of this code you're mentioning is just extracting $os_name, $arch, and $os_version, and setting $REMOTE_DOCKER_URL for windows image, which are still needed either way.

yliaog commented:
/retest

yliaog commented:
/lgtm

yliaog commented:
what is left to do to get the PR merged?

bclau commented:
> 
> 
> what is left to do to get the PR merged?

As far as this PR goes, nothing. But in the infra side - plenty. The idea is to also have Windows images build and published under the same gcr.io registry with the rest of the test images. For that, there are a few puzzle pieces missing from the overall picture:

1. Image promoter. Basically, the idea is to have the images built and published automatically without any human intervention into a staging repo, from which they will be promoted to the regular gcr.io registry used in tests. Right now, that image promoter exists, but it doesn't handle kubernetes/kubernetes images currently. For this, I have sent a couple of PRs that would help on that regard:

https://github.com/kubernetes/kubernetes/pull/84058
https://github.com/kubernetes/k8s.io/pull/400
https://github.com/kubernetes/test-infra/pull/14833

2. All images should be buildable. In this regard, there is an issue with the echoserver image that will have to be addressed first: https://github.com/kubernetes/kubernetes/issues/84158

3. Windows nodes will have to be included in the Image Promoter, so the Windows images can be built. Someone will have to give a hand on this part. Someone from Google was looking into adding Windows nodes to the Image Promoter, but I don't know about any follow ups. 

But if you need to build your own Windows images, you can already do so, you'll just have to push to a your own registry, like we do with the ``e2eteam`` registry. The "Windows Container images support" is split in 3 parts, and part 3 contains all 3 parts for simplicity, so you can just use that branch. Just read the ``README.md`` modified by this PR if you need to know how to create your own setup.

markyjackson-taulia commented:
Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this issue still intended for 1.17?

bclau commented:
> 
> 
> Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this issue still intended for 1.17?

All the necessary PRs have been sent. All they need is reviews.

markyjackson-taulia commented:
@bclau Just wanted to circle back on this. It appears to be stuck on some conflicts for 5 days. Just a kind reminder that code freeze is 11/14

bclau commented:
> 
> 
> @bclau Just wanted to circle back on this. It appears to be stuck on some conflicts for 5 days. Just a kind reminder that code freeze is 11/14

I didn't notice, sorry.

bclau commented:
/test pull-kubernetes-e2e-gce

bclau commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

markyjackson-taulia commented:
@bclau Checking back on this as we quickly approach the 11/14 code freeze

markyjackson-taulia commented:
@bclau @adelina-t Just a kind reminder that code freeze is 11/14. If are are not going to make this milestone let me know so I can move it to v1.18

bclau commented:
/hold cancel

markyjackson-taulia commented:
/milestone v1.18

markyjackson-taulia commented:
/milestone v1.18

claudiubelu commented:
/test pull-kubernetes-e2e-gce-100-performance

claudiubelu committed:
test images: Use multiple Windows nodes to build images

In order to build Windows container images for multiple OS versions,
--isolation=hyperv is required. However, not all clouds / nodes supports
or have it enabled by default, which is why we're going to rely on
having multiple nodes to build the Windows images, until this issue
is addressed.

claudiubelu committed:
Image Promoter: Adds Windows build nodes for Windows test images

We have added and enabled the Image Promoter on the k/k test images, which
will build the conformance images after a PR that affects kubernetes/test/images
merges.

We have added support for image-util.sh to handle external Windows Docker connections
in order to build Windows images.

This PR enables the Image Promoter to use some Windows nodes to build the necessary
Windows images.

claudiubelu commented:
/retest

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/test pull-kubernetes-e2e-gce-100-performance

claudiubelu commented:
/retest

claudiubelu commented:
/retest

claudiubelu commented:
/retest

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/retest

PatrickLang commented:
/retest

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/retest

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/test pull-kubernetes-e2e-gce

yliaog commented:
/lgtm

BenTheElder commented:
> A Windows node with Docker installed is required to build Windows images. The connection URL to it must be set in the REMOTE_DOCKER_URL_${os_version} env variable. Additionally, the authentication to the remote docker node is done through certificates, which must be found in ~/.docker-${os-version}.

I haven't had a chance to see how related this is but noting that I would actually like to see us remove the remote docker daemon stuff from the existing build as-is, it adds a lot of unnecessary complexity when the build could just be invoked on the remote host.. :/

In the future the dockerized build can preform pretty efficiently with native go caching without the rsync dance we do today.

I think some of the other build maintainers are in agreement here. the main reason it hasn't been dropped already is that removing it will be somewhat involved.

BenTheElder commented:
changing the image naming template is also a breaking change imo, I'm not sure we should be doing that. I'm certain there are consumers of the named images.

BenTheElder commented:
> A Windows node will be required though. I could share mine if needed.

this really needs to be something managed under wg-k8s-infra, it's not sustainable to have to borrow your node every time we need to push these (also the push is automated now anyhow ...?)

listx commented:
Today if I do `make all-push WHAT=test-webserver` then a bunch of images get built/pushed into `gcr.io/kubernetes-e2e-test-images`. Let's say `test-webserver` was pushed as version `1.0` in this case. If we want to now push Windows containers separately (that is, only build Windows images), is that possible (without overwriting the artifacts from the first `make all-push ...` invocation)? If yes, how does that play with manifest lists?

claudiubelu commented:
> 
> 
> > A Windows node with Docker installed is required to build Windows images. The connection URL to it must be set in the REMOTE_DOCKER_URL_${os_version} env variable. Additionally, the authentication to the remote docker node is done through certificates, which must be found in ~/.docker-${os-version}.
> 
> ~I haven't had a chance to see how related this is but noting that I would actually like to see us remove the remote docker daemon stuff from the existing build as-is, it adds a lot of unnecessary complexity when the build could just be invoked on the remote host.. :/~

So, the REMOTE_DOCKER_URL bits are not too big IMO. They were smaller originally when the idea was to simply have a single Windows node build the images for all other Windows versions using the ``--isolation=hyperv`` option. The multiple Windows nodes idea came from the fact that GCE does not have Hyper-V enabled for some reason, and thus, ``--isolation=hyperv`` is not possible, so there was no other option. The plan was to drop the multiple Windows nodes bits after they've addressed that.

As an additional note, it is a lot easier to build go Windows binaries on Linux than on Windows; I've had issues with this in the past. Plus, having the REMOTE_DOCKER_URL bits in ``image-util.sh`` means that it is easier to coordinate the image building workflow: in order to create the image manifest list for both Linux and Windows, all the images need to be build and pushed first, which is easier to ensure / control in the current state. And finally, now it's a single bash script, which is easier to maintain rather than having an additional PowerShell script for the image building on the Windows nodes (from my experience, people tend not to touch something like this).

> 
> ~In the future the dockerized build can preform pretty efficiently with native go caching without the rsync dance we do today.~
> 
> ~I think some of the other build maintainers are in agreement here. the main reason it hasn't been dropped already is that removing it will be somewhat involved.~
> 
> **EDIT**: The rsync / existing remote stuff is in the main makefile, not the test image build, ignore above.
> 
> Is it still possible to build for linux without needing a remote windows node? Ideally I should still be able to check out the repo and build and push my test images without relying on the pre-built ones or now needing to obtain a windows remote docker daemon.

Yes you can. Basically, the ``windows/amd64`` lines in the BASEIMAGE files are ignored if you don't have any ``REMOTE_DOCKER_URL`` env var set.

> 
> 
> changing the image naming template is also a breaking change imo, I'm not sure we should be doing that. Surely there are consumers of the named images?

Well, we can drop that if needed, even though I wouldn't be a fan of doing som Having soo many "images" for just one image feels bloat-y. And I don't know how is it breaking. Those are test images, so they should only be used for testing purposes. And additionally, we're not going to touch any of the already published images; they're still going to be there untouched. Plus, since we now have the Image Promoter, those new images are going to be published in a different registry.

> 
> 
> > A Windows node will be required though. I could share mine if needed.
> 
> this really needs to be something managed under wg-k8s-infra, it's not sustainable to have to borrow your node every time we need to push these (also the push is automated now anyhow ...?)

There is a resource group sponsored by Microsoft on Azure and under sig-windows, so it's not my node(s). And I've made changes to to this PR to allow the Image Promoter to use those Windows build nodes (see ``cloudbuild.yaml`` changes, and also: https://github.com/kubernetes/test-infra/pull/16083).

> 
> 
> Today if I do `make all-push WHAT=test-webserver` then a bunch of images get built/pushed into `gcr.io/kubernetes-e2e-test-images`. Let's say `test-webserver` was pushed as version `1.0` in this case. If we want to now push Windows containers separately (that is, only build Windows images), is that possible (without overwriting the artifacts from the first `make all-push ...` invocation)? If yes, how does that play with manifest lists?

No, not really. You can either build and push only for Linux images, or both Linux and Windows images. The only case in which you can build only Windows images is if the image's BASEIMAGE file only contains ``windows/amd64`` lines, and no linux ones (see the ``busybox/BASEIMAGE`` file added in this PR). If you reaally want to have only Windows images for the ``test-webserver`` image, clearly you can remove the Linux entries from its BASEIMAGE file. All the Linux images and tags will still be there, but the manifest list will be overwritten, because that's what the ``image-util.sh`` script does.

claudiubelu commented:
/retest

claudiubelu commented:
/test pull-kubernetes-e2e-kind

smourapina commented:
Hello @claudiubelu !
This is Bug Triage for release 1.18 checking in, with a friendly reminder that code freeze is coming in about 1 week (5 March), so please take this into account!

PatrickLang commented:
/lgtm
/approve

@smourapina - we're still blocked on reviews for e2e/testing. If you can help ping SIG-Testing too that would help

yliaog commented:
@listx is there any more concern about this PR?

listx commented:
/approve

PatrickLang commented:
/test pull-kubernetes-e2e-gce
